Update logic to compute the iOS localization paths to return full path

mokagio opened the pull request:
While running the code freeze for Simplenote 4.35, the lane failed on me with the following error:

```
	 7: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
	 6: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/bundler/gems/release-toolkit-413bdfd2a22e/lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_localize_project.rb:9:in `run'
	 5: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/bundler/gems/release-toolkit-413bdfd2a22e/lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:47:in `localize_project'
	 4: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/bundler/gems/release-toolkit-413bdfd2a22e/lib/fastlane/plugin/wpmreleasetoolkit/helper/git_helper.rb:81:in `commit'
	 3: from /Users/gio/.rbenv/versions/2.6.4/lib/ruby/2.6.0/forwardable.rb:230:in `sh'
	 2: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane/lib/fastlane/helper/sh_helper.rb:80:in `sh_control_output'
	 1: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'
/Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane_core/lib/fastlane_core/ui/interface.rb:153:in `shell_error!': [!] Exit status of command 'git add de.lproj/Localizable.strings he.lproj/Localizable.strings zh-Hant-TW.lproj/Localizable.strings ar.lproj/Localizable.strings el.lproj/Localizable.strings ja.lproj/Localizable.strings fa.lproj/Localizable.strings en.lproj/Localizable.strings en.lproj/InfoPlist.strings es.lproj/Localizable.strings it.lproj/Localizable.strings sv.lproj/Localizable.strings ko.lproj/Localizable.strings tr.lproj/Localizable.strings pt-BR.lproj/Localizable.strings ru.lproj/Localizable.strings cy.lproj/Localizable.strings fr.lproj/Localizable.strings id.lproj/Localizable.strings nl.lproj/Localizable.strings zh-Hans-CN.lproj/Localizable.strings' was 128 instead of 0. (FastlaneCore::Interface::FastlaneShellError)
fatal: pathspec 'de.lproj/Localizable.strings' did not match any files
```

Because we ran the glob in the localization folder, the strings path was relative to that folder, but we need it relative to the project root.

I used this branch to finish the code freeze and it worked. See [this commit](https://github.com/Automattic/simplenote-ios/pull/1251/commits/0c38c496671f05ab54fe782ff0c49e9bf0fac8da).

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:
```diff
@@ -41,9 +41,9 @@ def self.commit_version_bump(include_deliverfile: true, include_metadata: true)
-          strings_files = Dir.chdir(File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'])) do
-            Dir.glob('*.lproj/*.strings')
-          end
+          localizations_path = File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'])
+          strings_files = Dir.glob(File.join(localizations_path, '*.lproj/*.strings'))
```
You started using `File.join` for first part, but didn't use it for the `*.lproj/*.strings' part then ^^ 
```suggestion
          strings_files = Dir.glob(File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], '*.lproj', '*.strings'))
```

Note: slight risk with this solution is we open up to wildcard injection: if either of the `PROJECT_ROOT_FOLDER` or `PROJECT_NAME` env var contains a wildcard glob character (`*` or `?` typically), this part of the path too will end up being interpreted by `Dir.glob`. That's very likely why this was done via a `Dir.chrdir` before.

We can consider a low risk and very unlikely situation (honestly I'm good with it) and keep that. Or we can re-introduce `Dir.chdir` to avoid that risk 🤷 , something like this:
```suggestion
          loc_path = File.join(ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'])
          strings_files = Dir.chdir(loc_path) { Dir.glob('*.lproj/*.strings') }.map { |p| File.join(loc_path, p) }
```

Honestly I think the first solution is fine and simpler to read, and many ruby projects use `glob` that way. Just wanted to point out about the possible injection so that we're aware of it at least.

AliSoftware replied:
That being said, as pointed out in the PR review, in WC and WP those are not the path to the `.lproj` files either, so that won't probably fix it for them.

mokagio replied:
> You started using File.join for first part, but didn't use it for the `*.lproj/*.strings` part

🙃

jkmassel replied:
> Note: slight risk with this solution is we open up to wildcard injection: if either of the PROJECT_ROOT_FOLDER or PROJECT_NAME env var contains a wildcard glob character (* or ? typically), this part of the path too will end up being interpreted by Dir.glob. That's very likely why this was done via a Dir.chrdir before.

IMHO this is pretty unlikely, so I'd say just keep it clean for now 🤷‍♂️

AliSoftware reviewed (changes requested):
I'm a bit surprised that the `strings_files`  variable ever worked, in any project.

1. As you said, in the original code this returned relative paths, so that wouldn't work anyway
2. But even in the modified code, this will only find lproj files that are directly under the `PROJECT_NAME` folder. Which is not the case for all our apps

For example:
 - For `Simplenote`, the files seems indeed to be under `$PROJECT_ROOT_FOLDER/Simplenote/*.lproj`.
 - But for WooCommerce-iOS, they are under `$PROJECT_ROOT_FOLDER/WooCommerce/Resources/*.lproj`
 - For `WordPress`, the main strings files are also under `./WordPress/Resources/*.lproj`, but there are also `*.lproj` folders for widgets and extensions under `./WordPress/WordPressIntents/`, `./WordPress/WordPress*Extension/{Resources/}`, `./WordPress/*Widget/{Supporting Files/}`… do we want to commit them too? (depends what the `localize.py` script does, really)

So overall, I think this change would not fix apps other that Simplenote. Not sure why we never encountered the issue earlier, as the path was already wrong for all apps, but with this fix it will be correct for SN but still not for WP nor WC iinm.

mokagio commented:
@AliSoftware 

> So overall, I think this change would not fix apps other that Simplenote. Not sure why we never encountered the issue earlier, as the path was already wrong for all apps

🤔 As a matter of fact, I did not test the change outside of Simplenote... And I don't particularly like the use of env var either, but I just ~hacked~ updated the code to solve my immediate issue.

This looks like it needs more investigation... I'm going to revert it to draft. I don't have time to look into this just now, though.

Thanks for looking over this.

jkmassel commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:
```diff
@@ -41,9 +41,9 @@ def self.commit_version_bump(include_deliverfile: true, include_metadata: true)
         def self.localize_project
           Action.sh("cd #{ENV['PROJECT_ROOT_FOLDER']} && ./Scripts/localize.py")
```
While we're in here, WDYT about adding a check that these environment variables are present before starting this method? IMHO it would make anything that's missed far more obvious without having to dig into the source. Just thinking basically a runtime assertion (until we can refactor this to use proper arguments and stop assuming certain `ENV` keys exist, anyway...)

mokagio replied:
Oh yeah! I think we've been too relaxed with `ENV` vars and are paying the price now that we're changing things.

mokagio replied:
Partially addressed in 0239b8d. I didn't go down the full refactoring with DI route to prioritizie shipping the fix.

jkmassel reviewed (approved):
LGTM, but I left one potential improvement

mokagio committed:
DRY code fetching the paths of the iOS localization files for Git

mokagio committed:
Update logic to compute the iOS localization paths to return full path

While running the code freeze for Simplenote 4.35, the lane failed on me
with the following error:

```
         7: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
         6: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/bundler/gems/release-toolkit-413bdfd2a22e/lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_localize_project.rb:9:in `run'
         5: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/bundler/gems/release-toolkit-413bdfd2a22e/lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:47:in `localize_project'
         4: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/bundler/gems/release-toolkit-413bdfd2a22e/lib/fastlane/plugin/wpmreleasetoolkit/helper/git_helper.rb:81:in `commit'
         3: from /Users/gio/.rbenv/versions/2.6.4/lib/ruby/2.6.0/forwardable.rb:230:in `sh'
         2: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane/lib/fastlane/helper/sh_helper.rb:80:in `sh_control_output'
         1: from /Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane_core/lib/fastlane_core/ui/ui.rb:17:in `method_missing'
/Users/gio/Developer/a8c/snios/vendor/bundle/ruby/2.6.0/gems/fastlane-2.180.1/fastlane_core/lib/fastlane_core/ui/interface.rb:153:in `shell_error!': [!] Exit status of command 'git add de.lproj/Localizable.strings he.lproj/Localizable.strings zh-Hant-TW.lproj/Localizable.strings ar.lproj/Localizable.strings el.lproj/Localizable.strings ja.lproj/Localizable.strings fa.lproj/Localizable.strings en.lproj/Localizable.strings en.lproj/InfoPlist.strings es.lproj/Localizable.strings it.lproj/Localizable.strings sv.lproj/Localizable.strings ko.lproj/Localizable.strings tr.lproj/Localizable.strings pt-BR.lproj/Localizable.strings ru.lproj/Localizable.strings cy.lproj/Localizable.strings fr.lproj/Localizable.strings id.lproj/Localizable.strings nl.lproj/Localizable.strings zh-Hans-CN.lproj/Localizable.strings' was 128 instead of 0. (FastlaneCore::Interface::FastlaneShellError)
fatal: pathspec 'de.lproj/Localizable.strings' did not match any files
```

Because we ran the glob in the localization folder, the strings path was
relative to that folder, but we need it relative to the project root.

AliSoftware commented:
Encountered an issue that I think is related when I handled the WCiOS 6.5 submission today:

```
[17:49:46]: ---------------------------------
[17:49:46]: --- Step: ios_update_metadata ---
[17:49:46]: ---------------------------------
[17:49:46]: $ cd ./ && ./Scripts/update-translations.rb
[17:49:47]: ▸ Updating ar
[17:49:48]: ▸ WooCommerce/Resources/ar.lproj/Localizable.strings: OK
…

[17:50:11]: $ git commit -m Update\ translations
[17:50:11]: ▸ On branch release/6.5
[17:50:11]: ▸ Your branch is up to date with 'origin/release/6.5'.
[17:50:11]: ▸ Changes not staged for commit:
[17:50:11]: ▸ modified:   WooCommerce/Resources/ar.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/de.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/es.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/fr.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/he.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/id.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/it.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/ja.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/ko.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/nl.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/pt-BR.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/ru.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/sv.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/tr.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/zh-Hans.lproj/Localizable.strings
[17:50:11]: ▸ modified:   WooCommerce/Resources/zh-Hant.lproj/Localizable.strings
[17:50:11]: ▸ no changes added to commit
```
Which means that the translations didn't make it to the binary (and I only realized this once I submitted the build to Apple… so will have to redo another build with that now…)

This definitively needs to be fixed ASAP.

Maybe we can use some `Dir.glob('**/*.strings')` for the list of files to commit so that it words regardless of where the project has its strings file?

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:
```diff
@@ -58,15 +55,20 @@ def self.localize_project
+
+        def self.strings_files
+          Dir.glob(
+            File.join(
+              ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], '*.lproj', '*.strings'
+            )
```
I don't think this will work because I think `*` might be interpreted literally/verbatim here (not entirely sure if the shell substitution that would interpret that `*` as a wildcard comes into play here when we call the `commit` helper). Also, as mentioned previously, won't work for all the repos.

Maybe this instead would cover all cases?
```suggestion
            Dir.glob(File.join(
              ENV['PROJECT_ROOT_FOLDER'], ENV['PROJECT_NAME'], '**', '*.lproj', '*.strings'
            ))
```

jkmassel replied:
Interestingly this has a conflict on WPiOS, but it appears to work the way we want 🙂

I ran it locally and it fixed the issue.

mokagio replied:
@jkmassel what conflict are you seeing? 

I tried it just now (with @AliSoftware's `**` suggestion) and it seemed like it worked fine, as far as I can see. It definitely found and added the files: https://github.com/wordpress-mobile/WordPress-iOS/pull/16432.

jkmassel replied:
Just to update this - it was a conflict with some deprecated localizations – removed in https://github.com/wordpress-mobile/WordPress-iOS/pull/16415 and everything worked great after that.

AliSoftware commented:
Unsurprisingly, ran into it again during today's `code_freeze`. Just wanted to comment in order to highlight that it's not just happening during new betas and finalize release, but also during release branch cut… so practically during every step of our release process 😓  and indeed starts to become urgent to fix as this led to almost not shipping localizations (or submitting them into GlotPress in time for being translated) more than once, which wouldn't be pretty.

```
[15:19:53]: $ git commit -m Update\ strings\ for\ localization
[15:19:53]: ▸ On branch release/6.6
[15:19:53]: ▸ Your branch is up to date with 'origin/release/6.6'.
[15:19:53]: ▸ Changes not staged for commit:
[15:19:53]: ▸ modified:   WooCommerce/Resources/en.lproj/Localizable.strings
[15:19:53]: ▸ no changes added to commit
```

mokagio committed:
Make pattern looking for iOS localization file more flexible

Thanks @AliSoftware,
see
https://github.com/wordpress-mobile/release-toolkit/pull/236/files#r624029340

Using "**" should help us find the strings files regardless of where
they are located in the project file system.

mokagio committed:
Throw error if calling iOS localization commit without ENV values

It would have been better to refactor ios_git_helper.rb to not access
those values and use Dependency Injection instead. That can happen
later, right now, we want to ship the fix from this change set ASAP.

mokagio committed:
Refactor ios_git_helper.rb to leverage Hash methods

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:
```diff
@@ -56,17 +53,27 @@ def self.localize_project
+          Dir.glob(File.join(project_root, project_name, '**', '*.strings'))
+        end
+
+        def self.get_from_env!(key:)
+          ENV.fetch(key) do
+            UI.user_error! "Could not find value for \"#{key}\" in environment."
+          end
+        end
```
This is not the nicest method, and it's probably duplicated somewhere in the library.

As a followup, I think we may want to add this to `ENV` directly and find a nice name (`fetch!`?) for it and make it flexible.

AliSoftware replied:
How does the output look if you just use `ENV.fetch` without a block? According to the ruby documentation, `fetch` will raise a `KeyError` if you don't provide a block nor default value; wondering if Fastlane doesn't already catch that raised `KeyError` and prints a nice message for it already instead of you having to handle the `UI.user_error!` explicitly here?

AliSoftware replied:
(That being said I agree to ship this as is, and bike shed and circle back on this detail only later in separate PR if needs be)

mokagio commented on spec/ios_git_helper_spec.rb:
```diff
@@ -0,0 +1,21 @@
+require_relative './spec_helper'
+
+describe Fastlane::Helper::Ios::GitHelper do
```
Rubocop is telling me:

```
warning| RSpec/FilePath: Spec path should end with `fastlane/helper/ios/git_helper*_spec.rb`.
```

Something I'm not sure is why it's saying is for this spec but not for, say, `git_helper_spec.rb`. 🤔 

Anyways, if we were to change this, it would be appropriate to change all the others, too. Or, we can just add it to the excluded rules list. I'm more leaning towards the first option. Keen to hear what you think. Regardless, that should be a dedicated PR.

AliSoftware replied:
This is because it expects the module tree declared in this file to match the path subdirectories structure; so if the file contains a test for `describe Foo::Bar::Baz` then it expects the file to be in `Foo/bar/baz_spec.rb`

Indeed I can confirm that I've run into that one in the past and that this convention doesn't hold for many of our spec files in the current state of the codebase. I think most of them have been ignored in `rubocop-todo.yml` for now — because that's part of the many things we'd need to refactor quite a bit to fix all the violations and I haven't had time to fix them all. For this one I agree we'd need to align all the files to that convention at once, but [what do we say to rubocop spec violations…](https://66.media.tumblr.com/54876958f77d3a5e40d21f374ce31b68/tumblr_pqpq3mSIJ51rp74xfo2_400.gif)…

mokagio replied:
> I think most of them have been ignored in `rubocop-todo.yml` for now

Oh, right! I didn't think of looking in there 👍 

> but what do we say to rubocop spec violations…

🤣 (_still waiting for the last book!!!_)

mokagio reviewed (commented):
@jkmassel @AliSoftware here's what I consider a decent hotfix.

I manually tested it on [WordPress](https://github.com/wordpress-mobile/WordPress-iOS/pull/16432), [WooCommerce](https://github.com/woocommerce/woocommerce-ios/pull/4107), and [Simplenote](https://github.com/Automattic/simplenote-ios/pull/1269) (the latter, including the missing `ENV` values behavior). Feel free to have a go at it as well.

If this works as I think it does, I'd rather ship this as it is and followup to address architecture improvements or with things like extracting the `get_from_env!` method.

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_git_helper.rb:
```diff
@@ -56,17 +53,27 @@ def self.localize_project
         #
         def self.update_metadata
-          Action.sh("cd #{ENV['PROJECT_ROOT_FOLDER']} && ./Scripts/update-translations.rb")
+          Action.sh("cd #{get_from_env!(key: 'PROJECT_ROOT_FOLDER')} && ./Scripts/update-translations.rb")
```
Can't wait to get rid of the python script, of the odd requirement of having to `cd` to the right dir first for the script to work as expected, and for that script to be hosted in the app repo… 

![soon meme](https://i.kym-cdn.com/entries/icons/original/000/005/673/sooon.jpg)

AliSoftware reviewed (approved):
Code changes LGTM. Didn't actually test this from any of the host app repos yet (writing from my phone here) so maybe we should pick an app repo, point it to that branch and check it works as expected, but I'm pretty confident the logic is right (and we need this to move forward)
